.github/workflows/integration-tests.yml

```diff
@@ -0,0 +1,125 @@
+
+          ID="$(echo "$RESPONSE" | jq '.id')"
+          echo "Triggered variant analysis $ID"
+
+          if [ "$ID" == "null" ]; then
+            echo "Error triggering variant analysis"
+            exit 1
+          fi
+
+          echo "::set-output name=variant_anlaysis_id::$ID"
+
+      - name: Wait for variant analysis to complete
+        run: |
+          while true; do
+            RESPONSE=$(curl -H "Authorization: Bearer ${{ secrets.BOT_TOKEN }}" "https://api.github.com/repos/$GITHUB_REPOSITORY/code-scanning/codeql/variant-analyses/${{ steps.trigger.outputs.variant_anlaysis_id }}")
+            STATUS="$(echo "$RESPONSE" | jq '.status' -r)"
+            echo "Variant analysis ${{ steps.trigger.outputs.variant_anlaysis_id }} status: $STATUS"
+            if [ "$STATUS" == "completed" ]; then
+              echo "Exiting..."
+              exit 0
+            fi
+            sleep 10s
+          done
+
+      - name: Validate variant anlaysis status
+        id: validate
+        run: |
+          RESPONSE=$(curl -H "Authorization: Bearer ${{ secrets.BOT_TOKEN }}" "https://api.github.com/repos/$GITHUB_REPOSITORY/code-scanning/codeql/variant-analyses/${{ steps.trigger.outputs.variant_anlaysis_id }}")
+          echo "Response: $RESPONSE"
+
+          if [ "$(echo "$RESPONSE" | jq '.failure_reason')" != "null" ]; then
+            echo "Failure reason is not null"
+            exit 1
+          fi
+
+          if [ "$(echo "$RESPONSE" | jq '.scanned_repositories | length')" != "2" ]; then
+            echo "Number of scanned repos is not 2"
+            exit 1
+          fi
+
+          if [ "$(echo "$RESPONSE" | jq '.scanned_repositories[].repository.full_name' -r | sort)" != "$(echo -e "docker/compose\nhashicorp/terraform")" ]; then
+            echo "Full names of scanned repos is incorrect"
+            exit 1
+          fi
+
+          if [ "$(echo "$RESPONSE" | jq '.skipped_repositories.not_found_repos.repository_count')" != "1" ]; then
+            echo "Number of not found skipped repos is not 1"
+            exit 1
+          fi
+
+          if [ "$(echo "$RESPONSE" | jq '.skipped_repositories.not_found_repos.repository_full_names[]' -r)" != "github/does-not-exist" ]; then
+            echo "Not found skipped repos is incorrect"
+            exit 1
+          fi
+
+          if [ "$(echo "$RESPONSE" | jq '.skipped_repositories.access_mismatch_repos.repository_count')" != "0" ]; then
+            echo "Number of access mismatch skipped repos is not 0"
+            exit 1
+          fi
+
+          if [ "$(echo "$RESPONSE" | jq '.skipped_repositories.no_codeql_db_repos.repository_count')" != "0" ]; then
+            echo "Number of no CodeQL DB skipped repos is not 0"
+            exit 1
+          fi
+
+          if [ "$(echo "$RESPONSE" | jq '.skipped_repositories.over_limit_repos.repository_count')" != "0" ]; then
+            echo "Number of over limit skipped repos is not 0"
+            exit 1
+          fi
```

Comment:
All of these tests seem to be testing the GitHub API, rather than this action. This action shouldn't influence any of these, right? The only thing this action should influence is the status of individual repo tasks, which doesn't seem to be checked here.

Reply:
That's a good observation. I hadn't really considered it that way but you're right. I think this work has become a little conflated in my head with the idea of providing integration tests of all of variant analysis.

I'm now wondering if it makes sense to slim this test down to only consider the repo tasks, or whether to keep the current content and increase it further to cover the repo tasks. I'm leaning towards to second option as it's little extra effort and then it means these tests could be useful just to run on the `main` branch. Probably not possible / worth it for testing API changes before merging, but maybe just useful as a periodic sanity check.

Reply:
I agree that it would be really useful to have integration tests of the complete variant analysis flow, but I'm not sure that the correct place is within the action. Would it be possible to separate this out and think about implementing this as e.g. a scheduled workflow, independent of the action?

Reply:
I've added some checks of the `scanned_repositories` section that checks that both analyses succeeded, have the correct number of results, and have an artifact uploaded.

Regarding the question of whether these tests belong here or test the right thing, I think the situation is not perfect but keeping the test here is best for now. At least I'd prefer to implement these new tests here and then think about moving them later. I don't think we need another set of integration tests, so I'd prefer to either keep them here or move them elsewhere but not have this style of tests in two places.

Honestly, running these tests on every pull request of the action seems unnecessary. The only reason it makes sense having them in this repository is because this is the only repo where it's possible to test changes in this way before you merge them, but these integration tests are testing the API at least as much as they are testing this action. So I wouldn't be opposed to moving these tests out of this action and converting them to only run on schedule and on demand. Then when making changes to this action you rely on the unit tests, and can run the integration tests manually if you deem it necessary.

Does that fit with how you're thinking, and how much of this do you think we need to answer before merging this PR?

Reply:
I think it would still make sense to run these integration tests as part of this action since this repository contains essentially all the code that runs as part of the variant analysis (including the workflow that is triggered and the actions that are run). I'm not sure there's a better place for these integration tests either since that would just be duplicating part of the integration tests from this repository.

I'm happy to merge the PR without having an answer on this question since it will definitely improve the state of the integration tests and we can think about having integration tests for all of variant analysis later.